go/parquet/writer_properties.go

```diff
@@ -0,0 +1,510 @@
+type ColumnProperties struct {
+	Encoding          Encoding
+	Codec             compress.Compression
+	DictionaryEnabled bool
+	StatsEnabled      bool
+	MaxStatsSize      int64
+	CompressionLevel  int
+}
+
+// DefaultColumnProperties returns the default properties which get utilized for writing.
+//
+// The default column properties are the following constants:
+//	Encoding:						Encodings.Plain
+//	Codec:							compress.Codecs.Uncompressed
+//	DictionaryEnabled:	DefaultDictionaryEnabled
+//	StatsEnabled:				DefaultStatsEnabled
+//	MaxStatsSize:				DefaultMaxStatsSize
+//	CompressionLevel:		compress.DefaultCompressionLevel
+func DefaultColumnProperties() ColumnProperties {
+	return ColumnProperties{
+		Encoding:          Encodings.Plain,
+		Codec:             compress.Codecs.Uncompressed,
+		DictionaryEnabled: DefaultDictionaryEnabled,
+		StatsEnabled:      DefaultStatsEnabled,
+		MaxStatsSize:      DefaultMaxStatsSize,
+		CompressionLevel:  compress.DefaultCompressionLevel,
+	}
+}
+
+type writerPropConfig struct {
+	wr            *WriterProperties
+	encodings     map[string]Encoding
+	codecs        map[string]compress.Compression
+	compressLevel map[string]int
+	dictEnabled   map[string]bool
+	statsEnabled  map[string]bool
+}
+
+// WriterProperty is used as the options for building a writer properties instance
+type WriterProperty func(*writerPropConfig)
+
+// WithAllocator specifies the writer to use the given allocator
+func WithAllocator(mem memory.Allocator) WriterProperty {
+	return func(cfg *writerPropConfig) {
+		cfg.wr.mem = mem
+	}
+}
+
+// WithDictionaryDefault sets the default value for whether to enable dictionary encoding
+func WithDictionaryDefault(dict bool) WriterProperty {
+	return func(cfg *writerPropConfig) {
+		cfg.wr.defColumnProps.DictionaryEnabled = dict
+	}
+}
+
+// WithDictionaryFor allows enabling or disabling dictionary encoding for a given column path string
+func WithDictionaryFor(path string, dict bool) WriterProperty {
+	return func(cfg *writerPropConfig) {
+		cfg.dictEnabled[path] = true
+	}
+}
+
+// WithDictionaryPath is like WithDictionaryFor, but takes a ColumnPath type
+func WithDictionaryPath(path ColumnPath, dict bool) WriterProperty {
+	return WithDictionaryFor(path.String(), dict)
+}
+
+// WithDictionaryPageSizeLimit is the limit of the dictionary at which the writer
+// will fallback to plain encoding instead
+func WithDictionaryPageSizeLimit(limit int64) WriterProperty {
+	return func(cfg *writerPropConfig) {
+		cfg.wr.dictPagesize = limit
+	}
+}
+
+// WithBatchSize specifies the number of rows to use for batch writes to columns
+func WithBatchSize(batch int64) WriterProperty {
+	return func(cfg *writerPropConfig) {
+		cfg.wr.batchSize = batch
+	}
+}
+
+// WithMaxRowGroupLength specifies the number of rows as the maximum number of rows for a given row group in the writer.
+func WithMaxRowGroupLength(nrows int64) WriterProperty {
```

Comment:
not sure if this was mirrored on the C++ implementation intentionally, just wanted to check that this pattern is idiomatic go?

Reply:
The C++ implementation uses a Builder pattern that constructs an object with a bunch of "Set...." functions on it and then eventually calls "Build" to generate the resulting object. This follows the idiomatic Go pattern for taking options, which is the same pattern used in the Arrow Go library for things like the ipc.NewWriter etc.

Essentially it ends up looking like:

```go
parquet.NewWriterProperties(parquet.WithMaxRowGroupLength(nrows), parquet.With.........)
```

where `NewWriterProperties` can take any number of these `WriterProperty` options including none which would just produce the default WriterProperties.